xarray/core/indexes.py

```diff
@@ -45,72 +88,315 @@ def concat(
+            A new Index object.
+        """
         raise NotImplementedError(
             f"{self!r} doesn't support alignment with inner/outer join method"
         )
 
     def reindex_like(self: T_Index, other: T_Index) -> dict[Hashable, Any]:
+        """Query the index with another index of the same type.
+
+        Implementation is optional but required in order to support alignment.
+
+        Parameters
+        ----------
+        other : Index
+            The other Index object used to query this index.
+
+        Returns
+        -------
+        dim_positional_indexers : dict
+            A dictionary where keys are dimension names and values are positional
+            indexers.
+        """
         raise NotImplementedError(f"{self!r} doesn't support re-indexing labels")
 
-    def equals(self, other):  # pragma: no cover
+    def equals(self: T_Index, other: T_Index) -> bool:
+        """Compare this index with another index of the same type.
+
+        Implemenation is optional but required in order to support alignment.
+
+        Parameters
+        ----------
+        other : Index
+            The other Index object to compare with this object.
+
+        Returns
+        -------
+        is_equal : bool
+            ``True`` if the indexes are equal, ``False`` otherwise.
+        """
         raise NotImplementedError()
 
-    def roll(self, shifts: Mapping[Any, int]) -> Index | None:
+    def roll(self: T_Index, shifts: Mapping[Any, int]) -> T_Index | None:
+        """Roll this index by an offset along one or more dimensions.
+
+        This method can be re-implemented in subclasses of Index, e.g., when the
+        index can be itself indexed.
+
+        If not re-implemented, this method returns ``None``, i.e., calling
+        :py:meth:`Dataset.roll` will either drop the index in the resulting
+        dataset or pass it unchanged if its corresponding coordinate(s) are not
+        rolled.
+
+        Parameters
+        ----------
+        shifts : mapping of hashable to int, optional
+            A dict with keys matching dimensions and values given
+            by integers to rotate each of the given dimensions, as passed
+            :py:meth:`Dataset.roll`.
+
+        Returns
+        -------
+        rolled : Index
+            A new index with rolled data.
+        """
         return None
 
     def rename(
-        self, name_dict: Mapping[Any, Hashable], dims_dict: Mapping[Any, Hashable]
-    ) -> Index:
+        self: T_Index,
+        name_dict: Mapping[Any, Hashable],
+        dims_dict: Mapping[Any, Hashable],
+    ) -> T_Index:
+        """Maybe update the index with new coordinate and dimension names.
+
+        This method should be re-implemented in subclasses of Index if it has
+        attributes that depends on coordinate or dimension names.
+
+        By default (if not re-implemented), it returns the index itself.
+
+        Warning: the input names are not filtered for this index, they may
```

Comment:
```suggestion
        Warning: the input names are not filtered for this method, they may
```

I found this warning confusing. So for rename, we call `rename` on all indexes passing the dict ? And indexes are expected to return `self` if none of the names are relevant